discord/component.go

```diff
@@ -0,0 +1,93 @@
+package discord
+
+import "encoding/json"
+
+// ComponentType is the type of a component.
+type ComponentType uint
+
+const (
+	ActionBarComponent ComponentType = iota + 1
+	ButtonComponent
+)
+
+// Component is a component that can be attached to an interaction response.
+type Component interface {
+	Type() ComponentType
+	MarshalJSON() ([]byte, error)
+}
+
+// ActionBar is a row of components at the bottom of a message.
+type ActionBar struct {
+	Components []Component `json:"components"`
+}
+
+// Type implements the InteractionComponent interface.
+func (ActionBar) Type() ComponentType {
+	return ActionBarComponent
+}
+
+// MarshalJSON marshals the action bar in the format Discord expects.
+func (a ActionBar) MarshalJSON() ([]byte, error) {
+	type actionBar ActionBar
+
+	return json.Marshal(struct {
+		actionBar
+		Type ComponentType `json:"type"`
+	}{
+		actionBar: actionBar(a),
+		Type:      ActionBarComponent,
+	})
+}
+
+// Button is a clickable button that may be added to an interaction response.
+type Button struct {
+	Label string `json:"label"`
+	// CustomID attached to InteractionCreate event when clicked.
+	CustomID string       `json:"custom_id"`
+	Style    ButtonStyle  `json:"style"`
+	Emoji    *ButtonEmoji `json:"emoji,omitempty"`
+	// Present on link-style buttons.
+	URL      string `json:"url,omitempty"`
+	Disabled bool   `json:"disabled,omitempty"`
+}
+
+// ButtonStyle is the style to display a button in.
+type ButtonStyle uint
+
+// All types of ButtonStyle documented.
+const (
+	PrimaryButton   ButtonStyle = iota + 1 // Blurple button.
+	SecondaryButton                        // Grey button.
+	SuccessButton                          // Green button.
+	DangerButton                           // Red button.
+	LinkButton                             // Button that navigates to a URL.
+)
+
+// ButtonEmoji is the emoji displayed on the button before the text.
+type ButtonEmoji struct {
+	Name     string `json:"name"`
+	ID       string `json:"id"`
```

Comment:
Is there a reason this doesn't have a specific type?

Reply:
It has been updated to be a `discord.EmojiID`.